javascript/node/index.js

```diff
@@ -0,0 +1,102 @@
+import {
+    createRequest,
+    isEmptyMessage,
+    streamBlocks,
+    createAuthInterceptor,
+    createRegistry,
+    fetchSubstream
+} from '@substreams/core';
+import { createConnectTransport } from "@connectrpc/connect-node";
+import { writeCursor, getCursor } from "./cursor.js";
+import { isErrorUnresolvable } from "./error.js";
+
+const TOKEN = process.env.SUBSTREAMS_API_TOKEN
+const ENDPOINT = "https://mainnet.eth.streamingfast.io"
+const SPKG = "https://storage.googleapis.com/substreams-registry/spkg/ethereum-explorer-v0.1.1.spkg"
+const MODULE = "map_block_meta"
+const START_BLOCK = '100000'
+const STOP_BLOCK = '+10000'
+
+/*
+    Entrypoint of the application.
+    Because of the long-running connection, Substreams will disconnect from time to time.
+    The application MUST handle disconnections and commit the provided cursor to avoid missing information.
+*/
+const main = async () => {
+    const pkg = await fetchPackage()
+    const registry = createRegistry(pkg);
+
+    const transport = createConnectTransport({
+        baseUrl: ENDPOINT,
+        interceptors: [createAuthInterceptor(TOKEN)],
+        useBinaryFormat: true,
+        jsonOptions: {
+            typeRegistry: registry,
+        },
+    });
+
+    let running = true;
+    
+    // The infite loop handles disconnections. Every time an disconnection error is thrown, the loop will automatically reconnect
+    // and start consuming from the latest commited cursor.
+    while (running) {
+        try {
+            running = false;
+            await stream(pkg, registry, transport);
+        } catch (e) {
+            if (!isErrorUnresolvable(e)) {
+                running = true;
+            }
+
+            console.log(e)
+        }
+    }
+}
+
+const fetchPackage = async () => {
+    return await fetchSubstream(SPKG)
+}
+
+const unpack = async (response, registry) => {
+  if (response.message.case === "blockScopedData") {
+    const output = response.message.value.output?.mapOutput;
+    const cursor = response.message.value.cursor;
+
+    if (output !== undefined) {
+      const message = output.unpack(registry);
+      if (message === undefined) {
+        throw new Error(`Failed to unpack output of type ${output.typeUrl}`);
+      }
+
+      return {'output': output, 'cursor': cursor};
+    }
+  }
+
```

Comment:
We should "show" the important other messages that developers should handle so that it's at least a "bit" documented.

```
if (response.message.case === "blockUndoSignal") {
    // The blockchain you are streaming from undo 1 or more blocks and you must now handle that case.
    // The field `response.message.<last_valid_block>` contains the last valid block, you must undo whatever
    // has been done prior that (so for data where `block_number > last_valid_block`). Once undo, you must also
    // write the `response.message.<last_valid_cursor>`. In this example, we just print the undo signal and write the cursor.
    console.log(`Blockchain undo 1 or more blocks, returning to valid block #${<last_valid_block.num>} (${<last_valid_block.id>)`
    <write cursor>
}

if (response.message.case === "<find correct progress message value>") {
  // Progress messages are emitted each 2 second, we suggest logging one of such message each N seconds 
  // at your convenience with `JSON.stringify(response.message.value)`
} 
```

Reply:
We have to show that on undo, the `last_valid_cursor` must become the active cursor. Which makes me think about Helium and wondering if they got that corner case correctly.